src/duk_js_executor.c

```diff
@@ -2250,6 +2252,17 @@ DUK_INTERNAL void duk_js_execute_bytecode(duk_hthread *exec_thr) {
 		 * whenever a thread switch occurs by the DUK_HEAP_SWITCH_THREAD() macro.
 		 */
 #ifdef DUK_USE_INTERRUPT_COUNTER
+
+		if (thr->heap->dbg_processing) {
+			/* Don't interrupt if we're currently doing debug processing as
+			 * this may cause the debugger to be called recursively. Check required
+			 * for correct operation of throw intercept and other "exotic" halting
+			 * scenarios.
+			 */
+
+			goto skip_interrupt;
```

Comment:
I think that right now Eval also consumes the interrupt counter. It just doesn't execute the interrupt handler: https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L1709-L1733

As you see, the interrupt counter is updated during recursive interrupts but no actions are taken. The debug code that tries to validate that interrupt counter/init counter match bytecode execution count exactly is updated.

If Eval opcodes were not to be counted, they should also be ignored for the purposes of updating `thr->heap->inst_count_interrupt` in the main dispatch loop. Otherwise the two values won't match - that's not a concrete issue now, but the attempted guarantee is for them to match. I agree it might actually be preferable not to count Eval and other debugger induced bytecode execution towards a sampling profiler. But that's a separate change, and it'd be nice if there was one consistent policy in place at a time.

Anyway, it doesn't have to be fixed with this pull, we can handle these cleanups separately.